Debian #308825; don't close all descriptors before starting the slave.

XXX needs to be clarified and perhaps merged upstream

Index: gdm-2.18.1/daemon/display.c
===================================================================
--- gdm-2.18.1.orig/daemon/display.c	2007-05-02 18:38:48.000000000 +0200
+++ gdm-2.18.1/daemon/display.c	2007-05-02 18:41:44.000000000 +0200
@@ -306,8 +306,17 @@
 
 	closelog ();
 
-	/* Close everything */
-	gdm_close_all_descriptors (0 /* from */, fds[0] /* except */, slave_fifo_pipe_fd /* except2 */);
+	/* Debian changes */
+#if 0
+	/* upstream version */
+-	/* Close everything */
+-	gdm_close_all_descriptors (0 /* from */, fds[0] /* except */, slave_fifo_pipe_fd /* except2 */);
+#endif
+	/* Close stdin/stdout/stderr.  Leave others, as pam modules may have them open */
+	VE_IGNORE_EINTR (close (0));
+	VE_IGNORE_EINTR (close (1));
+	VE_IGNORE_EINTR (close (2));
+	/* End of Debian changes */
 
 	/* No error checking here - if it's messed the best response
          * is to ignore & try to continue */
Index: gdm-2.18.1/daemon/slave.c
===================================================================
--- gdm-2.18.1.orig/daemon/slave.c	2007-05-02 18:38:48.000000000 +0200
+++ gdm-2.18.1/daemon/slave.c	2007-05-02 18:42:54.000000000 +0200
@@ -1635,6 +1635,9 @@
 
 	closelog ();
 
+	/* Debian changes */
+#if 0
+	/* upstream version */
 	gdm_close_all_descriptors (0 /* from */, p[1] /* except */, -1 /* except2 */);
 
 	/* No error checking here - if it's messed the best response
@@ -1642,6 +1645,17 @@
 	gdm_open_dev_null (O_RDONLY); /* open stdin - fd 0 */
 	gdm_open_dev_null (O_RDWR); /* open stdout - fd 1 */
 	gdm_open_dev_null (O_RDWR); /* open stderr - fd 2 */
+#endif
+	/* Leave stderr open to the log */
+	VE_IGNORE_EINTR (close (0));
+	VE_IGNORE_EINTR (close (1));
+	gdm_close_all_descriptors (3 /* from */, p[1] /* except */, -1 /* except2 */);
+
+	/* No error checking here - if it's messed the best response
+         * is to ignore & try to continue */
+	gdm_open_dev_null (O_RDONLY); /* open stdin - fd 0 */
+	gdm_open_dev_null (O_RDWR); /* open stdout - fd 1 */
+	/* End of Debian changes */
 
 	openlog ("gdm", LOG_PID, LOG_DAEMON);
 
@@ -1797,6 +1811,9 @@
 
 		closelog ();
 
+		/* Debian changes */
+#if 0
+		/* upstream version */
 		gdm_close_all_descriptors (0 /* from */, slave_fifo_pipe_fd /* except */, d->slave_notify_fd /* except2 */);
 
 		/* No error checking here - if it's messed the best response
@@ -1804,6 +1821,17 @@
 		gdm_open_dev_null (O_RDONLY); /* open stdin - fd 0 */
 		gdm_open_dev_null (O_RDWR); /* open stdout - fd 1 */
 		gdm_open_dev_null (O_RDWR); /* open stderr - fd 2 */
+#endif
+		/* Leave stderr open to the log */
+		VE_IGNORE_EINTR (close (0));
+		VE_IGNORE_EINTR (close (1));
+		gdm_close_all_descriptors (3 /* from */, slave_fifo_pipe_fd /* except */, d->slave_notify_fd /* except2 */);
+
+		/* No error checking here - if it's messed the best response
+		 * is to ignore & try to continue */
+		gdm_open_dev_null (O_RDONLY); /* open stdin - fd 0 */
+		gdm_open_dev_null (O_RDWR); /* open stdout - fd 1 */
+		/* End of Debian changes */
 
 		openlog ("gdm", LOG_PID, LOG_DAEMON);
 
